### PR TITLE
【Prim】Add exp,multiply,expand vjp rules and Fix relative bugs 

### PR DESCRIPTION
**Pull request by cxxly**
### PR types

New features
### PR changes

Others
### Describe

- add exp, multiply, expand vjp rules 
- add exp, expand prim api
- fix infer shape bugs in _append_backward_ops
- fix target_gradients shape bugs @JiabinYang 
- fix run_program_op bugs @JiabinYang 
- add relative test cases

### CODE REVIEW

File: python/paddle/jit/dy2static/partial_program.py [hunk 1]
```diff
@@ -784,7 +782,7 @@ def _get_forward_backward_program_form(
     ):
         # NOTE(dev): We apply build_strategy for backward firstly to
         # avoid skipping more gc variables.
-        backward_start_op_index = forward_end_op_index + 2 * len(
+        backward_start_op_index = forward_end_op_index + 1 * len(
```
Comment (JiabinYang) [inline python/paddle/jit/dy2static/partial_program.py:780]: maybe need to change translated_layer.py, test_run_program_op, test_eager_run_program_op

File: python/paddle/fluid/backward.py [hunk 2]
```diff
@@ -1365,10 +1359,40 @@ def update_distop_context(
...
+        if core.is_prim_enabled():
+
+            def find_op_index(block_desc, cur_op_desc):
+                for idx in range(block_desc.op_size()):
+                    if cur_op_desc == block_desc.op(idx):
+                        return idx
+                return -1
+
+            grad_op_desc, op_grad_to_var = core.get_grad_op_desc(
+                composite_block.desc.op(find_op_index(block.desc, op.desc)),
```
Comment (JiabinYang) [inline python/paddle/fluid/backward.py:1386]: Where is code for coping op from sandbox back to original block

**Review by JiabinYang — COMMENTED** [review]
some comments

File: python/paddle/fluid/backward.py (hunk 2 above)
Comment (cxxly) [reply python/paddle/fluid/backward.py:1386]: It's unnecessary to copy here .  The line 1527 will add op_desc to original block.

File: python/paddle/jit/dy2static/partial_program.py (hunk 1 above)
Comment (cxxly) [reply python/paddle/jit/dy2static/partial_program.py:780]: DONE

File: python/paddle/fluid/backward.py [hunk 3]
```diff
@@ -1365,10 +1360,40 @@ def update_distop_context(
...
+        #       x_grad = reduce_sum(x_grad_unreduce)
+        if core.is_prim_enabled():
+
+            def find_op_index(block_desc, cur_op_desc):
+                for idx in range(block_desc.op_size()):
+                    if cur_op_desc == block_desc.op(idx):
+                        return idx
+                return -1
+
+            grad_op_desc, op_grad_to_var = core.get_grad_op_desc(
```
Comment (JiabinYang) [inline python/paddle/fluid/backward.py:1385]: maybe `grad_op_desc`, `op_grad_to_var` used out of scope?

**Review by JiabinYang — COMMENTED** [review]
one coment

**Review by JiabinYang — APPROVED** [review]
LGTM

File: python/paddle/fluid/backward.py (hunk 3 above)
Comment (cxxly) [reply python/paddle/fluid/backward.py:1385]: Only python `fuction`、`class`、`module` have scope. Variable in `if` and `for` statement have the same scope with out of those statement.  But init variable before those statement  is more safely as `if` may not be executed. Will fix this problem in a new pr. @cxxly

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request closed without merging**